blog/2023-07-24-pwa-tech-1.md/index.md

```diff
@@ -0,0 +1,474 @@
+
+```js
+/** Get all clients including uncontrolled, but only those within SW scope */
+export function getAllClientsInScope() {
+    // Include uncontrolled clients: necessary to know if there are multiple
+    // tabs open upon first SW installation
+    return self.clients
+        .matchAll({
+            includeUncontrolled: true,
+        })
+        .then((clientsList) =>
+            // Filter to just clients within this SW scope, because other clients
+            // on this domain but outside of SW scope are returned otherwise
+            clientsList.filter((client) =>
+                client.url.startsWith(self.registration.scope)
+            )
+        )
+}
+```
+
+The service worker uses the [`self.clients.matchAll()` API](https://developer.mozilla.org/en-US/docs/Web/API/Clients/matchAll) with the `includeUncontrolled` option, since some tabs may be uncontrolled the first time the service worker installs. Then, since that function returns every open client on this domain, even ones outside of the scope of the service worker's control, the resulting clients need to be filtered down to just the clients in scope. After the service worker gets the right clients list, it posts a message back to the client to report the clients info. Then, the `getClientsInfo()` method returns a promise that either resolves to the clients info or rejects with a failure reason.
+
+If there is a service worker waiting to take over (either the `”WAITING”` or `”FIRST_ACTIVATION”` conditions above), and there is only tab of the app open, the `PWALoadingBoundary` will apply the ready update by calling the [`useNewSW()` method](https://github.com/dhis2/app-platform/blob/a3490e03a2c2c4e706b5fad644d8f3beffc4a81a/pwa/src/offline-interface/offline-interface.js#L192-L219) on the Offline Interface. The method instructs the new service worker to take over: it detects if this new service worker is the first one that has installed for this app or an update to an existing service worker, then sends either a ‘claim clients’ message to a first-install service worker or a ‘skip waiting’ message to an updated service worker. Skipping waiting or claiming clients by the service worker both result in a `controllerchange` event in open clients, which triggers the event listener that the Offline Interface set up on `navigator.serviceWorker` to listen for that event (recall from the ["Registration of the service worker"](#registration-of-the-service-worker) section). The listener will then call `window.location.reload()` to reload the page so that the page can load under the control of the new service worker.
```

Comment:
```suggestion
If there is a service worker waiting to take over (either the `"WAITING"` or `"FIRST_ACTIVATION"` conditions above), and there is only one tab of the app open, the `PWALoadingBoundary` will apply the ready update by calling the [`useNewSW()` method](https://github.com/dhis2/app-platform/blob/a3490e03a2c2c4e706b5fad644d8f3beffc4a81a/pwa/src/offline-interface/offline-interface.js#L192-L219) on the Offline Interface. The method instructs the new service worker to take over: it detects if this new service worker is the first one that has installed for this app or an update to an existing service worker, then sends either a ‘claim clients’ message to a first-install service worker or a ‘skip waiting’ message to an updated service worker. Skipping waiting or claiming clients by the service worker both result in a `controllerchange` event in open clients, which triggers the event listener that the Offline Interface set up on `navigator.serviceWorker` to listen for that event (recall from the ["Registration of the service worker"](#registration-of-the-service-worker) section). The listener will then call `window.location.reload()` to reload the page so that the page can load under the control of the new service worker.
```

Reply:
I don't see a diff between these lines? 🤔

Reply:
I changed the quotes from `”` to `"`

Reply:
Ah gotcha, thanks 👍